Add support for searching information on clearly defined

nicarl opened the pull request:
### Summary of changes

This adds support for fetching information from ClearlyDefined as described in #299.
![Peek 2022-02-27 21-32](https://user-images.githubusercontent.com/26271333/155898849-04e2f854-7a4c-483e-b935-bf7cb70b35fb.gif)
### Context and reason for change

Motivation described in #299.

### How can the changes be tested

See screenshot + execute unit tests

Fix #299

nicarl commented on package.json:
```diff
@@ -12,6 +12,7 @@
     "@mui/lab": "^5.0.0-alpha.70",
     "@mui/material": "^5.4.3",
     "@mui/styles": "^5.4.2",
+    "axios": "^0.26.0",
```
I decided to add axios at it is more convenient than `fetch`

nicarl commented on package.json:
```diff
@@ -32,6 +34,7 @@
     "spdx-license-ids": "^3.0.11",
     "stream-json": "^1.7.4",
     "upath": "^2.0.1",
+    "url": "^0.11.0",
```
I had to add this because `webpack 5` does not support polyfill anymore. We should look into this together.

nicarl commented on src/Frontend/Components/PackageSearchPopup/ClearlyDefinedPackageCard.tsx:
```diff
@@ -0,0 +1,120 @@
+    textOverflow: 'ellipsis',
+    overflow: 'overflow',
+    whiteSpace: 'nowrap',
+    width: '25%',
```
There are still two problems with the css here.I was not able to solve these problems:

![css_hell2](https://user-images.githubusercontent.com/26271333/155898951-f708ab41-1709-4bbc-a073-6f86c27df304.png)
![css_hell_1](https://user-images.githubusercontent.com/26271333/155898953-696af803-2690-4008-b52e-e5ea369893e4.png)

nicarl replied:
One problem seems to be that `width` in percentage does not play along with ellipsis :-(

benedikt-richter commented on src/Frontend/Components/PackageSearchPopup/ClearlyDefinedPackageCard.tsx:
```diff
@@ -0,0 +1,130 @@
+import React, { ReactElement, useState } from 'react';
+import MuiTypography from '@mui/material/Typography';
+import { useQuery } from 'react-query';
+import { Alert as MuiAlert, Skeleton as MuiSkeleton } from '@mui/material';
+import { AccordionDetails as MuiAccordionDetails } from '@mui/material';
+import { AccordionSummary as MuiAccordionSummary } from '@mui/material';
+import { Accordion as MuiAccordion } from '@mui/material';
```
Imports should follow the pattern used in line 6

benedikt-richter replied:
also other occurences

benedikt-richter commented on src/Frontend/Components/PackageSearchPopup/PackageSearchPopup.tsx:
```diff
@@ -0,0 +1,148 @@
+    () => searchPackagesOnClearlyDefined(debouncedSearchTerm),
+    {
+      refetchOnWindowFocus: false,
+      enabled: !!currentSearchTerm,
```
(suggestion) `enabled: !!currentSearchTerm` -> `enabled: Boolean(currentSearchTerm)`

benedikt-richter commented on src/Frontend/Components/PackageSearchPopup/PackageSearchPopup.tsx:
```diff
@@ -0,0 +1,148 @@
+        autoFocus={false}
+      />
+      {isLoading ? (
+        <div className={classes.spinner}>
+          <MuiSpinner />
+        </div>
```
Would be nice to extract to use consistently throughout the app

benedikt-richter commented on src/Frontend/Components/PackageSearchPopup/PackageSearchPopup.tsx:
```diff
@@ -0,0 +1,148 @@
+          <MuiSpinner />
+        </div>
+      ) : isError ? (
+        <MuiAlert severity="error">
+          {`Failed while fetching data${
+            error instanceof Error ? `: ${error.message}` : ''
+          }`}
+        </MuiAlert>
```
Would be nice to extract to use consistently throughout the app

benedikt-richter commented on src/Frontend/Components/AttributionColumn/PackageSubPanel.tsx:
```diff
@@ -58,6 +65,17 @@ export function PackageSubPanel(props: PackageSubPanelProps): ReactElement {
+                dispatch(openPopup(PopupType.PackageSearchPopup));
+              }}
+              disabled={!props.isEditable}
+              icon={<SearchIcon aria-label={'Search packages icon'} />}
```
the icon should be styled like to other icons in the app. Would also be good to move it to Icons.tsx

benedikt-richter commented on src/Frontend/Components/AttributionColumn/PackageSubPanel.tsx:
```diff
@@ -58,6 +65,17 @@ export function PackageSubPanel(props: PackageSubPanelProps): ReactElement {
+              tooltipTitle="Search for package information"
+              placement="right"
+              onClick={(): void => {
+                dispatch(openPopup(PopupType.PackageSearchPopup));
```
so far the sub panel is not connected to the Redux store. However, not sure this callback should be provided as prop from parent

benedikt-richter commented:
Some observations:
* the search icon also works for empty name and version in attributions. Maybe disable it if there is no name?
* the search icon does not work for signals. However, this is not transparent to the user as visually it looks the same as in the case of attributions
* if I open the edit attribution popup and click on the search icon the search popup opens. When clicking on "cancel", I do not get back to the edit attribution popup
* in the case where no results are found in the search popup it would be nice to display a hint like "no results found". Curently, the user might suspect it's broken in this case

nicarl commented:
> Some observations:
> 
>     * the search icon also works for empty name and version in attributions. Maybe disable it if there is no name?
This should be fixed now. The field is only populated if there is a package name.

>     * the search icon does not work for signals. However, this is not transparent to the user as visually it looks the same as in the case of attributions

I changed the styling to make the icon grey if it not activated.

>     * if I open the edit attribution popup and click on the search icon the search popup opens. When clicking on "cancel", I do not get back to the edit attribution popup

This should be fixed now, the openPopup needs to be an array to "remember" the previously opened popups

>     * in the case where no results are found in the search popup it would be nice to display a hint like "no results found". Curently, the user might suspect it's broken in this case
I added the text :+1:

nicarl commented on src/Frontend/Components/PackageSearchPopup/ClearlyDefinedPackageCard.tsx:
```diff
@@ -0,0 +1,132 @@
+              )}
+            </MuiTypography>
+            <MuiTypography noWrap className={classes.text}>
+              {isLoading ? <MuiSkeleton width="60px" /> : data?.url ?? ''}
```
The `<MuiSkeleton />` detects the type of `MuiTypography` from the parent (https://mui.com/components/skeleton/#inferring-dimensions)

nicarl replied:
I decided to do as you suggested

nicarl commented:
I will squash the commits before merging.

benedikt-richter commented:
It seems the PURL is not updated after adding an entry from clearly defined

![Screenshot 2022-03-02 at 12 39 38](https://user-images.githubusercontent.com/46576389/156355998-516c7050-6b5a-4cdc-b106-c60d2e07ea33.png)

benedikt-richter commented:
The two icons could be a bit smaller

![Screenshot 2022-03-02 at 12 39 38](https://user-images.githubusercontent.com/46576389/156356166-6455dd9b-f5ba-4ac3-b32a-8035f47beea8.png)

benedikt-richter commented:
It might be good to exclude the namespace from the search term, while keeping the type

nicarl commented on src/Frontend/Components/AttributionColumn/attribution-column-helpers.ts:
```diff
@@ -226,12 +226,11 @@ export function usePurl(
 
   useEffect(() => {
     setTemporaryPurl(generatePurlFromPackageInfo(displayPackageInfo) || '');
-    // Dependency for displayPackageInfo is not included because otherwise displayed purl will be regenerated after every new input
```
Did we see any performance penalty for this?

benedikt-richter replied:
If you add  displaypackageInfo to the dependency array, the UI has some delay if you type fast.  It's not too bad but you'll recognize.

nicarl replied:
I think in this case we should debounce the input via the debounce hook I added in this PR. What do you think?

benedikt-richter replied:
That should work :)

benedikt-richter replied:
(optional if you want to look into it) btw we also debounce the input for file search. Maybe this could also use the same hook

nicarl replied:
I decided to use namespace, type and name in the dependency array. This should fix 90% of the problems. I think rethinking purl input is out of scope for this feature.

benedikt-richter commented:
New behaviour:
* open the edit popup on report view
* edit some field
* press "cancel"
* the unsaved changes popup is shown
* press "undo" or "save" there
* expected: there is no popup shown and you're back to report view
* reality: you're still on edit popup

nicarl commented:
> 

Should be fixed now

benedikt-richter reviewed (approved):
awesome 👍
